File: photoncore_prototype/src/clements.py
"""
Clements Decomposition for Unitary Matrices

Decomposes any NxN unitary matrix into a sequence of 2x2 rotations
that can be implemented on an MZI mesh.

Reference: Clements et al., "Optimal design for universal multiport
interferometers" Optica 3, 1460-1465 (2016)
"""

import numpy as np
from typing import Tuple, List
from numba import jit


class ClementsDecomposition:
    """
    Decompose unitary matrices into MZI mesh parameters.

    The Clements architecture is optimal in terms of optical depth,
    using exactly N(N-1)/2 beam splitters for an NxN unitary.
    """

    def __init__(self, n: int):
        """
        Initialize decomposition for NxN matrices.

        Args:
            n: Matrix dimension
        """
        self.n = n
        self.n_mzis = n * (n - 1) // 2

    def decompose(self, U: np.ndarray) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
        """
        Decompose unitary matrix U into MZI phases.

        Args:
            U: Unitary matrix of shape (n, n)

        Returns:
            thetas: Internal phases, shape (n_mzis,)
            phis: External phases, shape (n_mzis,)
            output_phases: Output phase shifters, shape (n,)
        """
        assert U.shape == (self.n, self.n), f"Expected ({self.n}, {self.n}), got {U.shape}"

        # Check unitarity
        if not np.allclose(U @ U.conj().T, np.eye(self.n), atol=1e-6):
            raise ValueError("Input matrix is not unitary")

        # Work on a copy
        T = U.copy().astype(np.complex128)

        thetas = []
        phis = []

        # Nullify off-diagonal elements column by column
        # Using the Clements procedure (different from Reck)

        for k in range(self.n - 1):
            # Process column k, nullifying from bottom to top
            if k % 2 == 0:
                # Even column: nullify bottom-up
                for m in range(self.n - 1, k, -1):
                    # Nullify T[m, k] using rotation on rows (m-1, m)
                    theta, phi = self._nullify_element(T, m - 1, m, k)
                    thetas.append(theta)
                    phis.append(phi)

                    # Apply rotation T_left
                    T = self._apply_rotation_left(T, m - 1, m, theta, phi)
            else:
                # Odd column: nullify top-down
                for m in range(k + 1, self.n):
                    # Nullify T[k, m] using rotation on columns (m-1, m)
                    theta, phi = self._nullify_element_right(T, k, m - 1, m)
                    thetas.append(theta)
                    phis.append(phi)

                    # Apply rotation T_right
                    T = self._apply_rotation_right(T, m - 1, m, theta, phi)

        # Remaining diagonal phases
        output_phases = np.angle(np.diag(T))

        return (
            np.array(thetas),
            np.array(phis),
            output_phases
        )

    def _nullify_element(self, T: np.ndarray, i: int, j: int, k: int) -> Tuple[float, float]:
        """
        Find rotation angles to nullify T[j, k] using rows i and j.

        Returns theta, phi for the MZI.
        """
        a = T[i, k]
        b = T[j, k]

        if np.abs(b) < 1e-12:
            return 0.0, 0.0

        r = np.sqrt(np.abs(a)**2 + np.abs(b)**2)
        theta = np.arctan2(np.abs(b), np.abs(a))
        phi = np.angle(a) - np.angle(b)

        return theta, phi

    def _nullify_element_right(self, T: np.ndarray, k: int, i: int, j: int) -> Tuple[float, float]:
        """
        Find rotation angles to nullify T[k, j] using columns i and j.
        """
        a = T[k, i]
        b = T[k, j]

        if np.abs(b) < 1e-12:
            return 0.0, 0.0

        r = np.sqrt(np.abs(a)**2 + np.abs(b)**2)
        theta = np.arctan2(np.abs(b), np.abs(a))
        phi = np.angle(a) - np.angle(b)

        return theta, phi

    @staticmethod
    def _apply_rotation_left(T: np.ndarray, i: int, j: int,
                              theta: float, phi: float) -> np.ndarray:
        """Apply 2x2 rotation to rows i and j from the left."""
        cos_t = np.cos(theta)
        sin_t = np.sin(theta)
        exp_phi = np.exp(1j * phi)

        # Rotation matrix (transposed for left multiplication)
        new_i = exp_phi.conj() * cos_t * T[i, :] + exp_phi.conj() * sin_t * T[j, :]
        new_j = -sin_t * T[i, :] + cos_t * T[j, :]

        T[i, :] = new_i
        T[j, :] = new_j

        return T

    @staticmethod
    def _apply_rotation_right(T: np.ndarray, i: int, j: int,
                               theta: float, phi: float) -> np.ndarray:
        """Apply 2x2 rotation to columns i and j from the right."""
        cos_t = np.cos(theta)
        sin_t = np.sin(theta)
        exp_phi = np.exp(1j * phi)

        new_i = cos_t * T[:, i] - sin_t * T[:, j]
        new_j = exp_phi * sin_t * T[:, i] + exp_phi * cos_t * T[:, j]

        T[:, i] = new_i
        T[:, j] = new_j

        return T

    def reconstruct(self, thetas: np.ndarray, phis: np.ndarray,
                    output_phases: np.ndarray) -> np.ndarray:
        """
        Reconstruct unitary matrix from MZI phases.

        Args:
            thetas: Internal phases
            phis: External phases
            output_phases: Output phase shifters

        Returns:
            Reconstructed unitary matrix
        """
        from mzi import MZIMesh

        mesh = MZIMesh(self.n)
        mesh.set_phases(thetas, phis, output_phases)
        return mesh.get_matrix()


def clements_decompose(U: np.ndarray) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
    """
    Convenience function to decompose a unitary matrix.

    Args:
        U: Unitary matrix

    Returns:
        (thetas, phis, output_phases)
    """
    n = U.shape[0]
    decomp = ClementsDecomposition(n)
    return decomp.decompose(U)


def clements_reconstruct(thetas: np.ndarray, phis: np.ndarray,
                         output_phases: np.ndarray) -> np.ndarray:
    """
    Convenience function to reconstruct matrix from phases.
    """
    n = len(output_phases)
    decomp = ClementsDecomposition(n)
    return decomp.reconstruct(thetas, phis, output_phases)


def random_unitary(n: int) -> np.ndarray:
    """
    Generate a random unitary matrix using QR decomposition.

    Args:
        n: Matrix dimension

    Returns:
        Random unitary matrix
    """
    # Random complex matrix
    Z = (np.random.randn(n, n) + 1j * np.random.randn(n, n)) / np.sqrt(2)

    # QR decomposition
    Q, R = np.linalg.qr(Z)

    # Ensure unique decomposition
    d = np.diag(R)
    ph = d / np.abs(d)
    Q = Q @ np.diag(ph)

    return Q


def decomposition_error(U: np.ndarray, thetas: np.ndarray, phis: np.ndarray,
                        output_phases: np.ndarray) -> float:
    """
    Compute reconstruction error for a decomposition.

    Returns Frobenius norm of difference.
    """
    U_reconstructed = clements_reconstruct(thetas, phis, output_phases)
    return np.linalg.norm(U - U_reconstructed, 'fro')


class SVDDecomposition:
    """
    Alternative decomposition using SVD.

    Any matrix M = U @ S @ V^H where U, V are unitary and S is diagonal.
    For photonic implementation: two MZI meshes with diagonal attenuators.
    """

    def __init__(self, n: int):
        self.n = n
        self.clements = ClementsDecomposition(n)

    def decompose(self, M: np.ndarray) -> dict:
        """
        Decompose general matrix using SVD.

        Args:
            M: Input matrix (not necessarily unitary)

        Returns:
            Dictionary with U_phases, singular_values, V_phases
        """
        U, S, Vh = np.linalg.svd(M)

        # Decompose U and V^H
        U_thetas, U_phis, U_out = self.clements.decompose(U)

        # V^H is unitary
        V = Vh.conj().T
        V_thetas, V_phis, V_out = self.clements.decompose(V)

        return {
            'U_thetas': U_thetas,
            'U_phis': U_phis,
            'U_output': U_out,
            'singular_values': S,
            'V_thetas': V_thetas,
            'V_phis': V_phis,
            'V_output': V_out
        }


if __name__ == "__main__":
    print("Testing Clements Decomposition...")

    # Test various sizes
    for n in [4, 8, 16]:
        print(f"\n--- Testing {n}x{n} ---")

        # Generate random unitary
        U = random_unitary(n)

        # Decompose
        decomp = ClementsDecomposition(n)
        thetas, phis, output_phases = decomp.decompose(U)

        print(f"Number of MZIs: {len(thetas)}")
        print(f"Expected: {n*(n-1)//2}")

        # Reconstruct
        U_reconstructed = decomp.reconstruct(thetas, phis, output_phases)

        # Check error
        error = np.linalg.norm(U - U_reconstructed, 'fro')
        print(f"Reconstruction error: {error:.2e}")

        # Check unitarity of reconstruction
        is_unitary = np.allclose(
            U_reconstructed @ U_reconstructed.conj().T,
            np.eye(n),
            atol=1e-6
        )
        print(f"Reconstructed is unitary: {is_unitary}")

        if error < 1e-6:
            print("✓ PASSED")
        else:
            print("✗ FAILED")

    print("\n✓ All Clements decomposition tests passed!")

File: photoncore_rust/src/clements.rs
//! Clements Decomposition for Unitary Matrices
//!
//! Decomposes any NxN unitary matrix into MZI mesh parameters.

use ndarray::{Array1, Array2};
use num_complex::Complex64;
use rand::Rng;
use std::f64::consts::PI;

use crate::mzi::MZIMesh;

/// Clements decomposition for unitary matrices.
pub struct ClementsDecomposition {
    pub n: usize,
    pub n_mzis: usize,
}

impl ClementsDecomposition {
    /// Create new decomposition for NxN matrices.
    pub fn new(n: usize) -> Self {
        let n_mzis = n * (n - 1) / 2;
        Self { n, n_mzis }
    }

    /// Decompose unitary matrix U into MZI phases.
    pub fn decompose(&self, u: &Array2<Complex64>) -> (Vec<f64>, Vec<f64>, Vec<f64>) {
        assert_eq!(u.shape(), &[self.n, self.n]);

        let mut t = u.clone();
        let mut thetas = Vec::new();
        let mut phis = Vec::new();

        // Nullify off-diagonal elements column by column
        for k in 0..(self.n - 1) {
            if k % 2 == 0 {
                // Even column: nullify bottom-up
                for m in (k + 1..self.n).rev() {
                    let (theta, phi) = self.nullify_element(&t, m - 1, m, k);
                    thetas.push(theta);
                    phis.push(phi);
                    self.apply_rotation_left(&mut t, m - 1, m, theta, phi);
                }
            } else {
                // Odd column: nullify top-down
                for m in (k + 1)..self.n {
                    let (theta, phi) = self.nullify_element_right(&t, k, m - 1, m);
                    thetas.push(theta);
                    phis.push(phi);
                    self.apply_rotation_right(&mut t, m - 1, m, theta, phi);
                }
            }
        }

        // Remaining diagonal phases
        let output_phases: Vec<f64> = (0..self.n).map(|i| t[[i, i]].arg()).collect();

        (thetas, phis, output_phases)
    }

    fn nullify_element(&self, t: &Array2<Complex64>, i: usize, j: usize, k: usize) -> (f64, f64) {
        let a = t[[i, k]];
        let b = t[[j, k]];

        if b.norm() < 1e-12 {
            return (0.0, 0.0);
        }

        let theta = b.norm().atan2(a.norm());
        let phi = a.arg() - b.arg();

        (theta, phi)
    }

    fn nullify_element_right(
        &self,
        t: &Array2<Complex64>,
        k: usize,
        i: usize,
        j: usize,
    ) -> (f64, f64) {
        let a = t[[k, i]];
        let b = t[[k, j]];

        if b.norm() < 1e-12 {
            return (0.0, 0.0);
        }

        let theta = b.norm().atan2(a.norm());
        let phi = a.arg() - b.arg();

        (theta, phi)
    }

    fn apply_rotation_left(
        &self,
        t: &mut Array2<Complex64>,
        i: usize,
        j: usize,
        theta: f64,
        phi: f64,
    ) {
        let cos_t = theta.cos();
        let sin_t = theta.sin();
        let exp_phi_conj = Complex64::from_polar(1.0, -phi);

        for col in 0..self.n {
            let ti = t[[i, col]];
            let tj = t[[j, col]];

            t[[i, col]] = exp_phi_conj * cos_t * ti + exp_phi_conj * sin_t * tj;
            t[[j, col]] = -sin_t * ti + cos_t * tj;
        }
    }

    fn apply_rotation_right(
        &self,
        t: &mut Array2<Complex64>,
        i: usize,
        j: usize,
        theta: f64,
        phi: f64,
    ) {
        let cos_t = theta.cos();
        let sin_t = theta.sin();
        let exp_phi = Complex64::from_polar(1.0, phi);

        for row in 0..self.n {
            let ti = t[[row, i]];
            let tj = t[[row, j]];

            t[[row, i]] = cos_t * ti - sin_t * tj;
            t[[row, j]] = exp_phi * sin_t * ti + exp_phi * cos_t * tj;
        }
    }

    /// Reconstruct unitary matrix from MZI phases.
    pub fn reconstruct(
        &self,
        thetas: &[f64],
        phis: &[f64],
        output_phases: &[f64],
    ) -> Array2<Complex64> {
        let mut mesh = MZIMesh::new(self.n, 0.0, 0.0, -100.0);
        mesh.set_phases(thetas, phis, output_phases);
        mesh.get_matrix(false)
    }
}

/// Generate a random unitary matrix using QR decomposition.
pub fn random_unitary(n: usize) -> Array2<Complex64> {
    let mut rng = rand::thread_rng();

    // Random complex matrix
    let mut z = Array2::zeros((n, n));
    for i in 0..n {
        for j in 0..n {
            z[[i, j]] = Complex64::new(
                rng.gen_range(-1.0..1.0),
                rng.gen_range(-1.0..1.0),
            ) / 2.0_f64.sqrt();
        }
    }

    // Simple Gram-Schmidt orthogonalization
    let mut q: Array2<Complex64> = Array2::zeros((n, n));

    for i in 0..n {
        let mut v = z.column(i).to_owned();

        // Subtract projections onto previous columns
        for j in 0..i {
            let q_j = q.column(j);
            let proj: Complex64 = q_j.iter().zip(v.iter()).map(|(a, b)| a.conj() * b).sum();
            for k in 0..n {
                v[k] -= proj * q[[k, j]];
            }
        }

        // Normalize
        let norm: f64 = v.iter().map(|x| x.norm_sqr()).sum::<f64>().sqrt();
        if norm > 1e-10 {
            for k in 0..n {
                q[[k, i]] = v[k] / norm;
            }
        }
    }

    q
}

/// Compute reconstruction error for a decomposition.
pub fn decomposition_error(
    u: &Array2<Complex64>,
    thetas: &[f64],
    phis: &[f64],
    output_phases: &[f64],
) -> f64 {
    let n = u.shape()[0];
    let decomp = ClementsDecomposition::new(n);
    let u_reconstructed = decomp.reconstruct(thetas, phis, output_phases);

    let diff = u - &u_reconstructed;
    diff.iter().map(|x| x.norm_sqr()).sum::<f64>().sqrt()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_decomposition_small() {
        let u = random_unitary(4);
        let decomp = ClementsDecomposition::new(4);

        let (thetas, phis, output) = decomp.decompose(&u);

        assert_eq!(thetas.len(), 6); // 4*3/2
        assert_eq!(phis.len(), 6);
        assert_eq!(output.len(), 4);

        let error = decomposition_error(&u, &thetas, &phis, &output);
        assert!(error < 1e-6, "Error too large: {}", error);
    }

    #[test]
    fn test_decomposition_medium() {
        let u = random_unitary(8);
        let decomp = ClementsDecomposition::new(8);

        let (thetas, phis, output) = decomp.decompose(&u);
        let error = decomposition_error(&u, &thetas, &phis, &output);

        assert!(error < 1e-5, "Error too large: {}", error);
    }

    #[test]
    fn test_random_unitary_is_unitary() {
        let u = random_unitary(8);
        let u_h = u.t().mapv(|x| x.conj());
        let product = u.dot(&u_h);

        for i in 0..8 {
            for j in 0..8 {
                let expected = if i == j { 1.0 } else { 0.0 };
                let diff = (product[[i, j]].norm() - expected).abs();
                assert!(diff < 1e-6, "Not unitary at [{},{}]: {}", i, j, diff);
            }
        }
    }
}
